### PR TITLE
Use cryptography for certificate generation

### PR DESCRIPTION
**Pull request by mhils**
Looking at #4368 turned up a mess of memory leaks in pyOpenSSL. However, [pyca/pyopenssl](https://github.com/pyca/pyopenssl) prominently states that "for anything other than making a TLS connection you should move to cryptography and drop your pyOpenSSL dependency." 

As such, instead of fixing pyOpenSSL, we just migrate and get rid of our problems. :) The cryptography API is much nicer, so we should use it everywhere where we currently use the pyOpenSSL objects.

@linw1995, does this fix the memory leaks you are seeing? 😃

### CODE REVIEW

**Comment by linw1995** [comment]
@mhils Yeah! It works well now. Thank you for the fantastic work 👍

**Comment by Prinzhorn** [comment]
When I read the title I was like "What did you use before using cryptography? Rocks and sticks? What's next? Use blockchain for certificate generation?". Now I understand the package is literally called cryptography :smile: 

Nice work!

File: mitmproxy/addons/tlsconfig.py [hunk 1]
```diff
@@ -126,17 +128,17 @@ def create_client_proxy_ssl_conn(self, tls_start: tls.TlsStartData) -> None:
...
+            chain_file=entry.chain_file,
             request_client_cert=False,
             alpn_select_callback=alpn_select_callback,
             extra_chain_certs=server.certificate_list,
             dhparams=self.certstore.dhparams,
         )
         tls_start.ssl_conn = SSL.Connection(ssl_ctx)
         tls_start.ssl_conn.set_app_data(AppData(
-            server_alpn=server.alpn,
+            server_alpn=always_bytes(server.alpn, "utf8", "replace"),
```
Comment (Kriechi) [inline mitmproxy/addons/tlsconfig.py:141]: Technically ALPN does not say anything about UTF-8 - just a plain byte string that is registered with IANA.
`replace` in this context sounds wrong/dangerous - but very unlikely to cause a problem in the real world.

File: mitmproxy/proxy/layers/http/__init__.py [hunk 2]
```diff
@@ -351,7 +351,7 @@ def check_killed(self, emit_error_hook: bool) -> layer.CommandGenerator[bool]:
                 yield HttpErrorHook(self.flow)
             # For HTTP/2 we only want to kill the specific stream, for HTTP/1 we want to kill the connection
             # *without* sending an HTTP response (that could be achieved by the user by setting flow.response).
-            if self.context.client.alpn == b"h2":
+            if self.context.client.alpn == "h2":
```
Comment (Kriechi) [inline mitmproxy/proxy/layers/http/__init__.py:354]: See above comment: technically this should be a byte string. There is no guarantee that bytes->utf8->bytes conversions will match again. Not only replace-encoding, but also different utf-8 code points etc.

File: mitmproxy/proxy/server.py [hunk 3]
```diff
@@ -427,7 +427,8 @@ def tls_start(tls_start: tls.TlsStartData):
                 tls_start.ssl_conn.set_accept_state()
             else:
                 tls_start.ssl_conn.set_connect_state()
-                tls_start.ssl_conn.set_tlsext_host_name(tls_start.context.client.sni)
+                if tls_start.context.client.sni is not None:
```
Comment (Kriechi) [inline mitmproxy/proxy/server.py:430]: Do we still support handshakes without SNI present?

File: mitmproxy/tools/console/flowdetailview.py [hunk 4]
```diff
@@ -40,64 +39,37 @@ def flowdetails(state, flow: mitmproxy.flow.Flow):
...
         )
 
-        c = sc.cert
+        c = sc.certificate_list[0]
         if c:
             text.append(urwid.Text([("head", "Server Certificate:")]))
             parts = [
                 ("Type", "%s, %s bits" % c.keyinfo),
-                ("SHA1 digest", c.digest("sha1")),
+                ("SHA256 digest", c.fingerprint().hex()),
```
Comment (Kriechi) [inline mitmproxy/tools/console/flowdetailview.py:63]: Nice!

File: mitmproxy/proxy/server.py (hunk 3 above)
Comment (mhils) [reply mitmproxy/proxy/server.py:430]: Yes, we do. The relevant logic is here: https://github.com/mitmproxy/mitmproxy/blob/7d7c2908b2b150e719e3f3c2f086e9290dc0603b/mitmproxy/addons/tlsconfig.py#L99-L106

**Review by Kriechi — APPROVED** [review]
Looks good - my only concern is that ALPN is *not* UTF-8. To be fully correct this should stay as byte strings - but I admit it is unlikely to cause problems in the real-world, unless somebody does crazy custom stuff.

File: mitmproxy/addons/tlsconfig.py (hunk 1 above)
Comment (mhils) [reply mitmproxy/addons/tlsconfig.py]: Fixed. 👍

**Pull request merged**